runtime/dynlink_nat.c

```diff
@@ -71,29 +71,42 @@ CAMLprim value caml_natdynlink_open(value filename, value global)
   void *sym;
   void *dlhandle;
   char_os *p;
-
-  /* TODO: dlclose in case of error... */
+  value exn;
 
   p = caml_stat_strdup_to_os(String_val(filename));
-  caml_enter_blocking_section();
+  exn = caml_enter_blocking_section_exn();
+  if (Is_exception_result(exn)) goto cleanup1;
   dlhandle = caml_dlopen(p, 1, Int_val(global));
-  caml_leave_blocking_section();
-  caml_stat_free(p);
+  exn = caml_leave_blocking_section_exn();
+  if (Is_exception_result(exn)) goto cleanup2;
 
-  if (NULL == dlhandle)
-    caml_failwith(caml_dlerror());
+  if (NULL == dlhandle) {
+    exn = caml_failwith_exn(caml_dlerror());
+    goto cleanup2;
+  }
```

Comment:
```suggestion
  if (NULL == dlhandle)
    caml_failwith(caml_dlerror());
```

Here, there is nothing to free. You can keep the original code.